### PR TITLE
Water Prison & Tavern

### PR DESCRIPTION
**Pull request by DjWarmonger**
- Handle Tavern and Prison on water
- Reveal FoW for newly released / recruited hero
- Fix the bug with inaccessible objects on water

### CODE REVIEW

**Comment by DjWarmonger** [comment]
Additionally I noticed that on minimap Prisons are colored gray, as if they were neutral heroes ¯\_(ツ)_/¯ But this is true also for land Prisons.

File: lib/CHeroHandler.cpp [hunk 1]
```diff
@@ -102,6 +102,17 @@ std::string CHero::getSpecialtyTooltipTextID() const
 	return TextIdentifier("hero", modScope, identifier, "specialty", "tooltip").get();
 }
 
+BoatId CHero::getBoatType() const
```
[the remark this answers is not in the document]
Comment (rilian-la-te) [reply lib/CHeroHandler.cpp:105]: Yes, agree. There is no reason to tie boat to hero class.

File: lib/mapObjects/CGHeroInstance.cpp [hunk 2]
```diff
@@ -437,28 +437,43 @@ void CGHeroInstance::onHeroVisit(const CGHeroInstance * h) const
...
+			
+			ObjectInstanceID boatId;
+			if (cb->gameState()->map->getTile(pos).isWater())
+			{
+				smp.val = maxMovePoints(false);
+				//Create a new boat for hero
+				NewObject no;
+				no.ID = Obj::BOAT;
+				no.subID = getBoatType().getNum();
+				no.pos = pos + int3(1,0,0); //FIXME: handle this magic offset somehow
```
[the remark this answers is not in the document]
Comment (DjWarmonger) [reply lib/mapObjects/CGHeroInstance.cpp:456]: This offset is specific to a boat (including the one bought at Shipyard or summoned) and has nothing to do with Prison offset.

File: lib/CHeroHandler.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (DjWarmonger) [reply lib/CHeroHandler.cpp:105]: Okay, I don't insist. Don't quite get how were these new boats intended to work.

File: lib/mapObjects/CGHeroInstance.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (DjWarmonger) [reply lib/mapObjects/CGHeroInstance.cpp]: Okay, moved it to static method of `CGBoat`.

**Review by IvanSavenko — APPROVED** [review]

**Pull request merged**